utxo/src/utxo_impl/utxo_storage.rs

```diff
@@ -0,0 +1,527 @@
+#![allow(dead_code, unused_variables, unused_imports)]
+// todo: remove ^ when all untested codes are tested
+
+use std::collections::{BTreeMap, HashMap};
+
+use crate::utxo_impl::{FlushableUtxoView, Utxo, UtxosCache, UtxosView};
+use crate::{BlockUndo, Error};
+use common::chain::block::Block;
+use common::chain::OutPoint;
+use common::primitives::{Id, H256};
+
+pub trait UtxosPersistentStorage {
+    fn set_utxo(&mut self, outpoint: &OutPoint, entry: Utxo) -> Result<(), crate::Error>;
+    fn del_utxo(&mut self, outpoint: &OutPoint) -> Result<(), crate::Error>;
+    fn get_utxo(&self, outpoint: &OutPoint) -> Result<Option<Utxo>, crate::Error>;
+    fn set_best_block_id(&mut self, block_id: &Id<Block>) -> Result<(), crate::Error>;
+    fn get_best_block_id(&self) -> Result<Option<Id<Block>>, crate::Error>;
+
+    fn set_undo_data(&mut self, id: Id<Block>, undo: &BlockUndo) -> Result<(), crate::Error>;
+    fn del_undo_data(&mut self, id: Id<Block>) -> Result<(), crate::Error>;
+    fn get_undo_data(&self, id: Id<Block>) -> Result<Option<BlockUndo>, crate::Error>;
+}
+
+pub struct UtxoDB<'a, S: UtxosPersistentStorage>(&'a mut S);
+
+impl<'a, S: UtxosPersistentStorage> UtxoDB<'a, S> {
+    pub fn new(store: &'a mut S) -> Self {
+        Self(store)
+    }
+
+    pub fn set_undo_data(&mut self, id: Id<Block>, undo: &BlockUndo) -> Result<(), crate::Error> {
+        self.0.set_undo_data(id, undo)
+    }
+
+    pub fn del_undo_data(&mut self, id: Id<Block>) -> Result<(), crate::Error> {
+        self.0.del_undo_data(id)
+    }
+
+    pub fn get_undo_data(&self, id: Id<Block>) -> Result<Option<BlockUndo>, crate::Error> {
+        self.0.get_undo_data(id)
+    }
+}
+
+impl<'a, S: UtxosPersistentStorage> UtxosView for UtxoDB<'a, S> {
+    fn get_utxo(&self, outpoint: &OutPoint) -> Option<Utxo> {
+        match self.0.get_utxo(outpoint) {
+            Ok(res) => res,
+            Err(_) => {
+                //todo: handle errors
```

Comment:
Should this return `Result<Option<Utxo>, Err>` instead?

Reply:
Fair point. I never was told why it was done like this.

Reply:
The thing is, while it makes sense for a database to do this, I can see that an error from the database is definitely irrecoverable. Besides, the UtxoView doesn't expect the storage to "fail". So a panic may be appropriate here.

Reply:
I'm fine with a panic if you think it's appropriate.

Reply:
I'll keep this unresolved because I need to think about it.